### PR TITLE
fsinfo: change to class-based style

### PR DESCRIPTION
**Pull request by allisonkarlitskaya**
Also, eject our self-rolled callback mechnism and change to the standard EventTarget API.

In order to keep type safety of the event types that we use, write a small wrapper, TypedEventTarget.  This is "mostly" safe, but we need a couple of casts: from the types on the standard EventTarget API, all event listener functions must support being called with all types of events, but ours only support being called with their specific event types.

This is starting to get ready for inclusion in the main cockpit library.

### CODE REVIEW

File: src/fsinfo.ts [hunk 1]
```diff
@@ -37,74 +58,102 @@ export interface FileInfo {
...
-    const callbacks: ((state: FileInfoState) => void)[] = [];
-
-    function close() {
-        channel.close();
+export class FsInfoCloseEvent extends TypedEvent<"close"> {
+    message: cockpit.JsonObject;
+
+    constructor(message: cockpit.JsonObject) {
+        super("close");
+        this.message = message;
```
Comment (cockpituous) [inline src/fsinfo.ts:97]: These 3 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 2]
```diff
@@ -37,74 +58,102 @@ export interface FileInfo {
...
+            this.partial_state = state as FsInfoBridgeState; // assume the bridge doesn't break protocol
+
+            if (is_json_dict(this.partial_state) && !this.partial_state.partial) {
+                this.state = { ...this.partial_state, loading: false };
+                this.dispatchEvent(new FsInfoChangeEvent(this.state));
+            }
+        });
+
+        this.channel.addEventListener("close", (_event, message) => {
+            this.dispatchEvent(new FsInfoCloseEvent(message));
```
Comment (cockpituous) [inline src/fsinfo.ts:134]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 3]
```diff
@@ -37,74 +58,102 @@ export interface FileInfo {
...
+        });
     }
 
-    function effect(callback: (state: FileInfoState) => void) {
-        callback(self.state);
-        callbacks.push(callback);
-        return () => callbacks.splice(callbacks.indexOf(callback), 1);
+    close() {
+        this.channel.close();
```
Comment (cockpituous) [inline src/fsinfo.ts:139]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 4]
```diff
@@ -37,74 +58,102 @@ export interface FileInfo {
...
     }
 
-    function entry(name: string): FileInfo | null {
-        return self.state.info?.entries?.[name] ?? null;
+    static entry(info: FileInfo, name: string): FileInfo | null {
+        return info.entries?.[name] ?? null;
```
Comment (cockpituous) [inline src/fsinfo.ts:143]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 5]
```diff
@@ -37,74 +58,102 @@ export interface FileInfo {
...
     }
 
-    function target(name: string): FileInfo | null {
-        const entries = self.state.info?.entries ?? {};
-        const targets = self.state.info?.targets ?? {};
+    static target(info: FileInfo, name: string): FileInfo | null {
+        const entries = info.entries ?? {};
+        const targets = info.targets ?? {};
```
Comment (cockpituous) [inline src/fsinfo.ts:148]: These 2 added lines are not executed by any test.

**Comment by allisonkarlitskaya** [comment]
I have to say I'm not crazy about this change.  The size of the code increased substantially on both the implementation side (all these event subtypes) and also the consumer side (separate add/remove calls and therefore losing the ability to do it inline)... and that's not even to mention the type-safe wrapper stuff (which honestly looks kinda scary)...

I agree that `EventTarget`/`addEventListener` and friends are more cockpit-flavoured, and are seemingly "the way" to do events in javascript, but this API really kinda sucks... and it's actually different from our own event-mixin which supports passing the "extra data" as arguments to the callback functions...

In short: I wonder if we should re-evaluate if we want to use some sort of other common third-party event library for standard Cockpit APIs in the future, or roll our own solution...

File: src/fsinfo.ts [hunk 6]
```diff
@@ -35,76 +56,78 @@ export interface FileInfo {
...
+            this.emit('close');
+        });
     }
 
-    function effect(callback: (state: FileInfoState) => void) {
-        callback(self.state);
-        callbacks.push(callback);
-        return () => callbacks.splice(callbacks.indexOf(callback), 1);
+    close() {
+        this.channel.close();
```
Comment (cockpituous) [inline src/fsinfo.ts:113]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 7]
```diff
@@ -35,76 +56,78 @@ export interface FileInfo {
...
     }
 
-    function entry(name: string): FileInfo | null {
-        return self.state.info?.entries?.[name] ?? null;
+    static entry(info: FileInfo, name: string): FileInfo | null {
+        return info.entries?.[name] ?? null;
```
Comment (cockpituous) [inline src/fsinfo.ts:117]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts [hunk 8]
```diff
@@ -35,76 +56,78 @@ export interface FileInfo {
...
     }
 
-    function target(name: string): FileInfo | null {
-        const entries = self.state.info?.entries ?? {};
-        const targets = self.state.info?.targets ?? {};
+    static target(info: FileInfo, name: string): FileInfo | null {
+        const entries = info.entries ?? {};
+        const targets = info.targets ?? {};
```
Comment (cockpituous) [inline src/fsinfo.ts:122]: These 2 added lines are not executed by any test.

**Comment by allisonkarlitskaya** [comment]
This looks a tonne better: https://github.com/cockpit-project/cockpit-files/compare/e1edf63899d808fb8a9e8f986fb7a11420f6907d..ba1b70f67b98a1d76ddd3f3a5c9e3f80278fb6ef

I want a review from Martin since my intent is that this will go upstream into the cockpit repo and become "the way" we do events in our "new" APIs...

File: src/fsinfo.ts (hunk 7 above)
Comment (cockpituous) [inline src/fsinfo.ts:117]: These 2 added lines are not executed by any test.

File: src/fsinfo.ts (hunk 8 above)
Comment (cockpituous) [inline src/fsinfo.ts:122]: These 2 added lines are not executed by any test.

**Comment by martinpitt** [comment]
This [rawhide failure](https://artifacts.dev.testing-farm.io/a9946b95-e0db-42b4-8a78-486db1eb8857/) is highly sus:
```
> log: PackageKit went away from D-Bus
head: cannot open '/home/admin/newdir/newfile' for reading: No such file or directory
Traceback (most recent call last):
  File "/source/test/check-application", line 1042, in testCopyPaste
    self.assertEqual(m.execute("head -n 1 /home/admin/newdir/newfile"), "test_text\n")
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

It could be that it wasn't from this PR, but already landed in main, during the time when rawhide tests were broken due to other crap? But the above is most definitively our bug.

File: src/fsinfo.ts [hunk 9]
```diff
@@ -35,76 +56,78 @@ export interface FileInfo {
...
     mtime?: number;
-    content?: string;
     target?: string;
-    entries?: {
-        [filename: string]: FileInfo;
-    };
-    targets?: {
-        [filename: string]: FileInfo;
-    };
+    entries?: Record<string, FileInfo>;
```
Comment (martinpitt) [inline src/fsinfo.ts:60]: Note to self: https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeys-type

File: src/fsinfo.ts [hunk 10]
```diff
@@ -113,33 +136,17 @@ export function fsinfo(path: string, attrs: string[], options?: cockpit.JsonObje
...
-            self.state = {
-                info: is_json_dict(state.info) ? state.info : null,
-                error: is_json_dict(state.error) ? state.error : null,
-                loading: false
-            };
-
-            for (const callback of callbacks) {
-                callback(self.state);
+export function fsinfo(path: string, attrs: (keyof FileInfo)[], options?: cockpit.JsonObject): Promise<FileInfo> {
+    return new Promise((resolve, reject) => {
```
Comment (martinpitt) [inline src/fsinfo.ts:142]: This looks like a backwards compat function, but the code in app.tsx was already ported to use `FsInfoClient` directly. This is also rather shallow -- do we need this for anything outside of c-files?

**Review by martinpitt — COMMENTED** [review]
Thanks! Some question/possible cleanup above, and I can't say for sure if the rawhide regression is introduced in this PR. If not, then we'll need a followup soon, as that's a release blocker.

File: src/fsinfo.ts (hunk 10 above)
Comment (allisonkarlitskaya) [reply src/fsinfo.ts:142]: This is not backcompat but is intended as a non-compatible API that does something different: a one-shot (watch: false) query of the current state, which it simply returns, or an exception.

I started writing a Qunit test for this on a branch in the cockpit repo...

Comment (martinpitt) [reply src/fsinfo.ts:142]: I see, ok! It's fine to keep it here then, I was just curious.

**Review by martinpitt — APPROVED** [review]
I retried the rawhide failure, just in case it is a flake. We need to fix this ASAP, though.

**Comment by martinpitt** [comment]
rawhide passed in #557, so it's not completely broken in main. Let's see what the retry says.

**Comment by martinpitt** [comment]
Retry worked! :+1: So it's a flake.

**Comment by jelly** [comment]
Great to see this merged I have some follow up questions:

* Should we now merge this back into Cockpit and finalize this as API? Thinking about my use fsinfo PR
* Secondly should we start using this new event class for to replace our own cockpit.js eventemitter?

**Comment by allisonkarlitskaya** [comment]
> Great to see this merged I have some follow up questions:
> 
>     * Should we now merge this back into Cockpit and finalize this as API? Thinking about my use fsinfo PR
> 
>     * Secondly should we start using this new event class for to replace our own cockpit.js eventemitter?

Answer to both questions is yes.

I have a branch I've started locally to get fsinfo upstream as a separate file in the `pkg/lib/` directory, under qunit tests.  I hope to push that soon.
